instrumentation/opentelemetry-instrumentation-remoulade/src/opentelemetry/instrumentation/remoulade/__init__.py

```diff
@@ -0,0 +1,135 @@
+from opentelemetry.semconv.trace import SpanAttributes
+
+_MESSAGE_TAG_KEY = "remoulade.action"
+_MESSAGE_SEND = "send"
+_MESSAGE_RUN = "run"
+
+_MESSAGE_NAME_KEY = "remoulade.actor_name"
+
+
+class InstrumentationMiddleware(Middleware):
+    def __init__(self, _tracer):
+        self._tracer = _tracer
+        self._span_registry = {}
+
+    def before_process_message(self, _broker, message):
+        trace_ctx = extract(message.options["trace_ctx"])
+        retry_count = message.options.get("retries")
+
+        operation_name = (
+            "remoulade/process"
+            if retry_count is None
+            else f"remoulade/process(retry-{retry_count})"
+        )
+
+        span = self._tracer.start_span(
+            operation_name, kind=trace.SpanKind.CONSUMER, context=trace_ctx
+        )
+
+        if retry_count is not None:
+            span.set_attribute("retry_count", retry_count)
+
+        activation = trace.use_span(span, end_on_exit=True)
+        activation.__enter__()
+
+        utils.attach_span(
+            self._span_registry, message.message_id, (span, activation)
+        )
+
+    def after_process_message(
+        self, _broker, message, *, result=None, exception=None
+    ):
+        span, activation = utils.retrieve_span(
+            self._span_registry, message.message_id
+        )
+
+        if span is None:
+            # no existing span found for message_id
+            return
+
+        if span.is_recording():
+            span.set_attribute(_MESSAGE_TAG_KEY, _MESSAGE_RUN)
+            span.set_attribute(_MESSAGE_NAME_KEY, message.actor_name)
+            pass
+
+        activation.__exit__(None, None, None)
+        utils.detach_span(self._span_registry, message.message_id)
+
+    def before_enqueue(self, _broker, message, delay):
+        retry_count = message.options.get("retries")
+
+        operation_name = (
+            "remoulade/send"
+            if retry_count is None
+            else f"remoulade/send(retry-{retry_count})"
+        )
+
+        span = self._tracer.start_span(
+            operation_name, kind=trace.SpanKind.PRODUCER
+        )
+
+        if retry_count is not None:
+            span.set_attribute("retry_count", retry_count)
+
+        if span.is_recording():
+            span.set_attribute(_MESSAGE_TAG_KEY, _MESSAGE_SEND)
+            span.set_attribute(
+                SpanAttributes.MESSAGING_MESSAGE_ID, message.message_id
+            )
+            span.set_attribute(_MESSAGE_NAME_KEY, message.actor_name)
+            pass
+
+        activation = trace.use_span(span, end_on_exit=True)
+        activation.__enter__()
+
+        utils.attach_span(
+            self._span_registry,
+            message.message_id,
+            (span, activation),
+            is_publish=True,
+        )
+
+        if "trace_ctx" not in message.options:
+            message.options["trace_ctx"] = {}
+        inject(message.options["trace_ctx"])
```

Comment:
If possible, please align with the way other instrumentations propagate over rabbitMQ messages. You can use [Pika](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation/opentelemetry-instrumentation-pika) for example

Reply:
Hey, I think I wasn't clear with my comment. What I meant was that pika instrumentation uses the headers of the request and not the options. This means a message sent using Pika and received using remoulade will not propagate context.

Reply:
The way remoulade enqueues messages and consumes them is not compatible with any other task processing over rabbitMQ libraries. For example, all information about a task is sent in the body of the rabbitMQ message and the headers are not used nor accessible. remoulade should always be used both for the producer and consumer parts.